crates/wasi-threads/src/lib.rs

```diff
@@ -0,0 +1,138 @@
+//! Implement [`wasi-threads`].
+//!
+//! [`wasi-threads`]: https://github.com/WebAssembly/wasi-threads
+
+use anyhow::{anyhow, Result};
+use rand::Rng;
+use std::thread;
+use wasi_common::I32Exit;
+use wasmtime::{Caller, Linker, Module, SharedMemory, Store};
+
+// This name is a function export designated by the wasi-threads specification:
+// https://github.com/WebAssembly/wasi-threads/#detailed-design-discussion
+const WASI_ENTRY_POINT: &str = "wasi_thread_start";
+
+pub struct WasiThreadsCtx<T> {
+    module: Module,
+    linker: Linker<T>,
+}
+
+impl<T: Clone + Send + 'static> WasiThreadsCtx<T> {
+    pub fn new(module: Module, linker: Linker<T>) -> Self {
+        Self { module, linker }
+    }
+
+    pub fn spawn(&self, host: T, thread_start_arg: i32) -> Result<i32> {
+        let module = self.module.clone();
+        let linker = self.linker.clone();
+
+        // Start a Rust thread running a new instance of the current module.
+        let wasi_thread_id = random_thread_id();
+        let builder = thread::Builder::new().name(format!("wasi-thread-{}", wasi_thread_id));
+        builder.spawn(move || {
+            // Convenience function for printing failures, since the `Thread`
+            // has no way to report a failure to the outer context.
+            let fail = |msg: String| {
+                format!(
+                    "wasi-thread-{} exited unsuccessfully: {}",
+                    wasi_thread_id, msg
+                )
+            };
+
+            // Each new instance is created in its own store.
+            let mut store = Store::new(&module.engine(), host);
+            let instance = linker
+                .instantiate(&mut store, &module)
+                .expect(&fail("failed to instantiate".into()));
+            let thread_entry_point = instance
+                .get_typed_func::<(i32, i32), ()>(&mut store, WASI_ENTRY_POINT)
+                .expect(&fail(format!(
+                    "failed to find wasi-threads entry point function: {}",
+                    WASI_ENTRY_POINT
+                )));
```

Comment:
I feel like this is a little tricky. Panicking on errors here doesn't feel like the right thing to do since these panics can theoreteically get triggered. In that sense I think that the errors need to be handled here. As you've probably noticed, however, there's nowhere to communicate the error to. One option would be to move the initialization here to the thread calling spawn, but that feels sort of wrong since it sort of belongs in the child thread.

Overall I think this might be best left as an issue on the `wasi-threads` repository. Basically how would a thread spawn operation get notified that the thread was spawned but something else failed? Or otherwise how are traps handled and/or other fatal errors in the threads proposal? (that is the `panic!` below also needs better handling I think)

Reply:
Yeah, I've looked into this more since you reviewed this. There has been some discussion [in a wasi-threads issue](https://github.com/WebAssembly/wasi-threads/pull/17/files) which boils down to "any `proc_exit` or trap in any thread exits the program." I was looking at [what `run.rs` does](https://github.com/bytecodealliance/wasmtime/blob/main/src/commands/run.rs#L215-L246) for traps and I think we would want to do something similar here--essentially call `std::process::exit`. The problem is that calling `std::process::exit` is likely not what one wants to do in an embedding scenario. I would suggest we do implement wasi-threads to initially `std::process::exit` and then open an issue to think about this for the embedding scenario: I suspect users here would want a trap or `proc_exit` to kill all the threads but not kill the original program calling Wasmtime but, with no great way to cancel threads in Rust, I think this might be kind of difficult.

Reply:
I'm ok with this all being a TODO and not necessarily covered here. At the very least this PR is a good step forward to being able to spawn threads.

I do think these issues need to be resolved before stabilization, but perhaps it's best to track these sorts of things in the wasi-threads repository?

Reply:
Ok, to resolve this I rebased again on `main` and applied [2ba9f11](https://github.com/bytecodealliance/wasmtime/pull/5484/commits/2ba9f11b56adaef32f5bc2596cafab8328e52ae4).